lms/envs/common.py

```diff
@@ -1197,7 +1197,9 @@ def _make_mako_template_dirs(settings):
 # once the responsibility of XBlock creation is moved out of modulestore - cpennington
 XBLOCK_MIXINS = (LmsBlockMixin, InheritanceMixin, XModuleMixin, EditInfoMixin)
 
-# Allow any XBlock in the LMS
+# .. setting_name: XBLOCK_SELECT_FUNCTION
+# .. setting_default: prefer_xmodules
+# .. setting_description: Function to select XBlock when multiple matching entrypoints.
```

Comment:
1. Are there a list of functions that are valid for this? Should we say what the options are, and what the meaning is behind each, or point someone to where they could learn about them?
2. I'm not clear on what an "XBlock entrypoint" means. Is there a little more that could be added to help clarify?  Can an example be provided?

Reply:
sure I'll try to come up with an example

Reply:
Note: #1 above I think is more important.  #2 would help some like me who doesn't know much about this, but if this would be clear to someone who knows XBlocks, maybe it isn't needed?  Thank you.

Reply:
I guess you didn't want to reference PRs `#1` and `#2` before :sweat_smile:

I also don't really understand this too much, but from what I understand, this settings variable is not really something configurable, as it needs to reference a function (not even a module path) that already exists in the edx codebase. The only other example I was able to find was [xblock's `default_select`](https://github.com/edx/XBlock/blob/master/xblock/plugin.py#L31).

from what I was researching, this is just a way to circle around a problem that could happen with [Python Package Entrypoints](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points) where basically an entrypoint group could define multiple entrypoints with the same name, so we somehow need to decide which one to pick.

a good example could be this: https://github.com/edx/edx-platform/blob/9e38b172025ede19120f038cae47b27beb6c3ce1/common/lib/xmodule/setup.py

where basically for the entrypoint group `xblock.v1`, we are including entrypoints for xmodules and xblocks. If somehow they were defining the same entrypoint (and entrypoint with the same `name`, let's say `about` was inside `XMODULES` and `XBLOCKS`), `prefer_xmodules` makes sure we use the one defined in `XMODULES` instead of `XBLOCKS`.

@robrap I hope this was a proper example, could you please help me with the wording now for this variable?

Reply:
The response is helpful. Some follow-up questions:
1. How does default_select work?
2. Maybe a description something like:
```
Function to select XBlock when multiple matching entrypoints.

Choose between prefer_xmodules, which will include all XBlocks in the LMS, but favors xmodules when their are duplicates, or default_select, which will favor the XBlock over the xmodule.

For an example of these XBlock entrypoints, see:
https://github.com/edx/edx-platform/blob/9e38b172025ede19120f038cae47b27beb6c3ce1/common/lib/xmodule/setup.py
```

Note: I’m not sure if what I wrote is even accurate?

Reply:
1. `default_select` just chooses the first one :smile: `prefer_xmodules` is a wrapper on top of it, making sure it chooses the `xmodule` ones, if there are still duplicates, it uses `default_select` which again, chooses the first one.
2. I don't think the second paragraph is correct, and this is an internal settings variable btw, so I don't think users will be able to customize this. I am not even sure if linking that example would be even helpful, linking them to the proper [python package EntryPoints explanation](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points) would be more helpful in my opinion, what do you think?

Reply:
Thank you. Here's a possible update.  If it is inaccurate, feel free to simply provide an updated suggestion:
```
Function used to select an XBlock from the python package EntryPoints.

Choose between prefer_xmodules and default_select. The function prefer_modules will choose the first xmodule if there is one, otherwise, it will act like default_select. The function default_select will simply choose the first match found.
```

Reply:
yep, that's perfect, thanks